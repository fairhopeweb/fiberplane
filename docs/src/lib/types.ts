import type { OpenAPI, OpenAPIV2, OpenAPIV3 } from "openapi-types";

export type Operation = Omit<OpenAPIV3.OperationObject, "tags"> & {
  httpVerb: HttpMethods;
  path: string;
};

export type TagObject = OpenAPIV3.TagObject;

export type ExtendedTagObject = TagObject & {
  operations: Operation[] | undefined;
};

export type PathItemObject = OpenAPIV3.PathItemObject;

export type OperationObject = OpenAPIV3.OperationObject;

export type HttpMethods = OpenAPIV3.HttpMethods;
