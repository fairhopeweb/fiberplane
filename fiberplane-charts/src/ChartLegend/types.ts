import type { Timeseries } from "../providerTypes";

export type { Timeseries };

export type ChartLegendProps = {
    /**
     * Handler that is invoked when the user toggles the visibility of a
     * timeseries.
     *
     * If no handler is specified, no UI for toggling the visibility of timeseries
     * is presented.
     */
    onToggleTimeseriesVisibility?: (event: ToggleTimeseriesEvent) => void;

    /**
     * Whether the chart is read-only.
     *
     * Set to `true` to disable interactive controls.
     */
    readOnly?: boolean;

    /**
     * Array of timeseries data to display in the legend.
     */
    timeseriesData: Array<Timeseries>;
};

export type ToggleTimeseriesEvent = {
    /**
     * The timeseries that was toggled.
     */
    timeseries: Timeseries;

    /**
     * If `true`, the visibility should be toggled of all timeseries *except* the
     * one specified.
     */
    toggleOthers: boolean;
};
