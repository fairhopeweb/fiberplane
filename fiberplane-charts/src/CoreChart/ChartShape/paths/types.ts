/**
 * Produces a coordinate value from a point.
 */
export type CoordinateFactory<P> = (p: P) => number;
