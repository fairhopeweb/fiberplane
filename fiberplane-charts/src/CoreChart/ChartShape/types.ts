import type { Scales } from "../types";

export type CommonShapeProps = {
  /**
   * Whether any shape list is focused.
   *
   * Some shapes indicate whether they are focused by de-emphasizing other
   * shapes. Such de-emphasizing should only be done when any (other) shape
   * is focused.
   */
  anyFocused: boolean;

  /**
   * Show a gradient under Line or Area shapes in the chart.
   */
  areaGradientShown: boolean;

  /**
   * The shape's color, based on the shape list it belongs to.
   */
  color: string;

  /**
   * Whether the shape list the shape belongs to is currently focused.
   */
  focused: boolean;

  /**
   * Chart scales for converting abstract coordinates to pixel values.
   */
  scales: Scales;
};
