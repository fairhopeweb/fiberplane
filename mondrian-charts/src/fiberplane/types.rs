use crate::types::*;
pub use fiberplane_models::notebooks::{GraphType, StackingType};
pub use fiberplane_models::providers::{Metric, ProviderEvent, Timeseries};
pub use fiberplane_models::timestamps::{TimeRange, Timestamp};
use serde::Serialize;
use std::collections::BTreeMap;

/// All the data necessary to generate an abstract chart from an array of
/// timeseries.
pub struct TimeseriesSourceData<'source, 'slice, 'value> {
    /// The type of chart to display.
    pub graph_type: GraphType,

    /// The type of stacking to apply to the chart.
    pub stacking_type: StackingType,

    /// Array of timeseries data to display in the chart.
    ///
    /// Make sure the timeseries contains data for the given time range, or you
    /// may not see any results.
    pub timeseries_data: &'slice [&'source Timeseries],

    /// The time range to be displayed.
    pub time_range: TimeRange,

    /// Additional values that will be plotted on the chart, but which are not
    /// part of any timeseries. These are not plotted, but are taken into
    /// account when deciding the range of the Y axis.
    pub additional_values: &'value [f64],
}

/// All the data necessary to generate an abstract chart from a combination of
/// timeseries data, events and an optional target latency.
pub struct CombinedSourceData<'source, 'slice> {
    /// The type of stacking to apply to the chart.
    ///
    /// **Warning:** This property is accepted for consistency, but setting it
    /// to anything except [GraphType::Line] will cause the events to be
    /// ignored.
    pub graph_type: GraphType,

    /// The type of stacking to apply to the chart.
    pub stacking_type: StackingType,

    /// Array of timeseries data to display in the chart.
    ///
    /// Make sure the timeseries contains data for the given time range, or you
    /// may not see any results.
    pub timeseries_data: &'slice [&'source Timeseries],

    /// Array of events to display in the chart.
    ///
    /// Note that events will not be displayed if the `graph_type` is anything
    /// other than [`GraphType::Line`].
    pub events: &'slice [&'source ProviderEvent],

    /// Optional target latency to display on the chart, in seconds.
    pub target_latency: Option<f64>,

    /// The time range to be displayed.
    pub time_range: TimeRange,
}

/// Source type for series in charts that contain combined data sources.
#[derive(Serialize)]
#[non_exhaustive]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum SeriesSource<'source> {
    Timeseries(&'source Timeseries),
    Events,
    TargetLatency,
}

impl<'source> From<&'source Timeseries> for SeriesSource<'source> {
    fn from(value: &'source Timeseries) -> Self {
        Self::Timeseries(value)
    }
}

/// Source type for points in charts that contain combined data sources.
#[derive(Serialize)]
#[non_exhaustive]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum PointSource<'source> {
    Metric(&'source Metric),
    Event(&'source ProviderEvent),
    None,
}

impl<'source> From<&'source Metric> for PointSource<'source> {
    fn from(value: &'source Metric) -> Self {
        Self::Metric(value)
    }
}

impl<'source> From<MondrianChart<&'source Timeseries, &'source Metric>>
    for MondrianChart<SeriesSource<'source>, PointSource<'source>>
{
    fn from(value: MondrianChart<&'source Timeseries, &'source Metric>) -> Self {
        Self {
            shape_lists: value.shape_lists.into_iter().map(ShapeList::into).collect(),
            x_axis: value.x_axis,
            y_axis: value.y_axis,
        }
    }
}

/// A map that holds arbitrary values per timestamp.
///
/// The keys in the map represent timestamps, while the values are generic and
/// are calculated over all the metrics matching that timestamp.
pub(crate) type Buckets<T> = BTreeMap<Timestamp, T>;

/// Represents the minimum and maximum values inside a series of numbers.
#[derive(Clone, Copy)]
pub(crate) struct MinMax(pub f64, pub f64);

impl MinMax {
    pub fn from_value(value: f64) -> Self {
        Self(value, value)
    }

    #[must_use]
    pub fn extend_with_other(mut self, other: Self) -> Self {
        if other.0 < self.0 {
            self.0 = other.0;
        }
        if other.1 > self.1 {
            self.1 = other.1;
        }

        self
    }

    #[must_use]
    pub fn extend_with_value(mut self, value: f64) -> Self {
        if value < self.0 {
            self.0 = value;
        } else if value > self.1 {
            self.1 = value;
        }

        self
    }
}

impl<'source> From<ShapeList<&'source Timeseries, &'source Metric>>
    for ShapeList<SeriesSource<'source>, PointSource<'source>>
{
    fn from(value: ShapeList<&'source Timeseries, &'source Metric>) -> Self {
        Self {
            shapes: value.shapes.into_iter().map(Shape::into).collect(),
            source: value.source.into(),
        }
    }
}

impl<'source> From<Shape<&'source Metric>> for Shape<PointSource<'source>> {
    fn from(other: Shape<&'source Metric>) -> Self {
        match other {
            Shape::Area(area) => Shape::Area(area.into()),
            Shape::Line(line) => Shape::Line(line.into()),
            Shape::Point(point) => Shape::Point(point.into()),
            Shape::Rectangle(rectangle) => Shape::Rectangle(rectangle.into()),
        }
    }
}

impl<'source> From<Area<&'source Metric>> for Area<PointSource<'source>> {
    fn from(value: Area<&'source Metric>) -> Self {
        Self {
            area_gradient_shown: None,
            points: value.points.into_iter().map(AreaPoint::into).collect(),
        }
    }
}

impl<'source> From<AreaPoint<&'source Metric>> for AreaPoint<PointSource<'source>> {
    fn from(value: AreaPoint<&'source Metric>) -> Self {
        Self {
            x: value.x,
            y_min: value.y_min,
            y_max: value.y_max,
            source: value.source.into(),
        }
    }
}

impl<'source> From<Line<&'source Metric>> for Line<PointSource<'source>> {
    fn from(value: Line<&'source Metric>) -> Self {
        Self {
            area_gradient_shown: None,
            points: value.points.into_iter().map(Point::into).collect(),
        }
    }
}

impl<'source> From<Point<&'source Metric>> for Point<PointSource<'source>> {
    fn from(value: Point<&'source Metric>) -> Self {
        Self {
            x: value.x,
            y: value.y,
            source: value.source.into(),
        }
    }
}

impl<'source> From<Rectangle<&'source Metric>> for Rectangle<PointSource<'source>> {
    fn from(value: Rectangle<&'source Metric>) -> Self {
        Self {
            x: value.x,
            y: value.y,
            width: value.width,
            height: value.height,
            source: value.source.into(),
        }
    }
}

pub(crate) type StackedChartBuckets = Buckets<StackedChartBucketValue>;

#[derive(Clone, Copy, Default)]
pub(crate) struct StackedChartBucketValue {
    /// Used to keep track of how much a bucket is "filled" while calculating
    /// the area shapes.
    pub current_y: f64,

    /// The sum of all values in the bucket.
    pub total: f64,
}
