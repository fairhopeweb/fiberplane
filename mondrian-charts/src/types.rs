pub use fiberplane_models::notebooks::{GraphType, StackingType};
pub use fiberplane_models::providers::{Metric, ProviderEvent, Timeseries};
pub use fiberplane_models::timestamps::{TimeRange, Timestamp};
use serde::Serialize;
use std::collections::BTreeMap;

/// All the data necessary to generate an abstract chart from an array of
/// timeseries.
pub struct TimeseriesSourceData<'source, 'slice> {
    /// The type of chart to display.
    pub graph_type: GraphType,

    /// The type of stacking to apply to the chart.
    pub stacking_type: StackingType,

    /// Array of timeseries data to display in the chart.
    ///
    /// Make sure the timeseries contains data for the given time range, or you
    /// may not see any results.
    pub timeseries_data: &'slice [&'source Timeseries],

    /// The time range to be displayed.
    pub time_range: TimeRange,
}

/// All the data necessary to generate an abstract chart from an array of
/// timeseries and an array of events.
///
/// Note we only support generating line charts from combined timeseries and
/// events data. If `graph_type` is anything other than [`GraphType::Line`],
/// the events will be ignored.
pub struct TimeseriesAndEventsSourceData<'source, 'slice> {
    /// The type of stacking to apply to the chart.
    ///
    /// **Warning:** This property is accepted for consistency, but setting it
    /// to anything except [GraphType::Line] will cause the events to be
    /// ignored.
    pub graph_type: GraphType,

    /// The type of stacking to apply to the chart.
    pub stacking_type: StackingType,

    /// Array of timeseries data to display in the chart.
    ///
    /// Make sure the timeseries contains data for the given time range, or you
    /// may not see any results.
    pub timeseries_data: &'slice [&'source Timeseries],

    /// Array of events to display in the chart.
    pub events: &'slice [&'source ProviderEvent],

    /// The time range to be displayed.
    pub time_range: TimeRange,
}

/// Source type for series in charts that contain combined data sources.
#[derive(Serialize)]
#[non_exhaustive]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum SeriesSource<'source> {
    Timeseries(&'source Timeseries),
    Events,
}

impl<'source> From<&'source Timeseries> for SeriesSource<'source> {
    fn from(value: &'source Timeseries) -> Self {
        Self::Timeseries(value)
    }
}

/// Source type for points in charts that contain combined data sources.
#[derive(Serialize)]
#[non_exhaustive]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum PointSource<'source> {
    Metric(&'source Metric),
    Event(&'source ProviderEvent),
}

impl<'source> From<&'source Metric> for PointSource<'source> {
    fn from(value: &'source Metric) -> Self {
        Self::Metric(value)
    }
}

/// An abstract chart with information about what to render and where to render
/// it.
///
/// All coordinates in an abstract chart are normalized to run from 0.0 to 1.0,
/// so (0, 0) is the origin of the chart (typically rendered bottom left), while
/// (1, 0) is the end of the X axis and (0, 1) is the end of the Y axis.
///
/// The generic argument `S` refers to the type of the series from which shapes
/// will be generated, while the type `P` refers to the type for individual data
/// points. When generating charts from timeseries data, these will be
/// [Timeseries] and [Metric], respectively.
#[derive(Default, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct AbstractChart<S, P> {
    pub x_axis: Axis,
    pub y_axis: Axis,
    pub shape_lists: Vec<ShapeList<S, P>>,
}

impl<'source> From<AbstractChart<&'source Timeseries, &'source Metric>>
    for AbstractChart<SeriesSource<'source>, PointSource<'source>>
{
    fn from(value: AbstractChart<&'source Timeseries, &'source Metric>) -> Self {
        Self {
            shape_lists: value.shape_lists.into_iter().map(ShapeList::into).collect(),
            x_axis: value.x_axis,
            y_axis: value.y_axis,
        }
    }
}

/// Defines the range of values that are displayed along a given axis.
#[derive(Default, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Axis {
    /// The value to display at the chart origin.
    pub min_value: f64,

    /// The value to display at the end of the axis.
    pub max_value: f64,

    /// Optional suggestion of where to draw ticks based on the detected bucket
    /// intervals.
    ///
    /// Ticks are expressed as values between `minValue` and `maxValue`.
    pub tick_suggestions: Option<Vec<f64>>,
}

impl Axis {
    /// Extends the range of the axis with the given interval.
    ///
    /// The range of the interval is divided among ends of the axis. This can be
    /// used to extend the axis with enough space to display the bars for the
    /// first and last buckets displayed on a bar chart.
    #[must_use]
    pub fn extend_with_interval(mut self, interval: f64) -> Self {
        let half_interval = 0.5 * interval;
        self.min_value -= half_interval;
        self.max_value += half_interval;
        self
    }
}

/// A map that holds arbitrary values per timestamp.
///
/// The keys in the map represent timestamps, while the values are generic and
/// are calculated over all the metrics matching that timestamp.
pub(crate) type Buckets<T> = BTreeMap<Timestamp, T>;

/// Represents the minimum and maximum values inside a series of numbers.
#[derive(Clone, Copy)]
pub(crate) struct MinMax(pub f64, pub f64);

impl MinMax {
    pub fn from_value(value: f64) -> Self {
        Self(value, value)
    }

    #[must_use]
    pub fn extend_with_other(mut self, other: Self) -> Self {
        if other.0 < self.0 {
            self.0 = other.0;
        }
        if other.1 > self.1 {
            self.1 = other.1;
        }

        self
    }

    #[must_use]
    pub fn extend_with_value(mut self, value: f64) -> Self {
        if value < self.0 {
            self.0 = value;
        } else if value > self.1 {
            self.1 = value;
        }

        self
    }
}

/// List of shapes that belongs together.
///
/// These should be rendered in the same color.
#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ShapeList<S, P> {
    pub shapes: Vec<Shape<P>>,

    /// The original source this shape list belongs to.
    ///
    /// This would be the type of input data the chart was generated from, such
    /// as [Timeseries].
    pub source: S,
}

impl<'source> From<ShapeList<&'source Timeseries, &'source Metric>>
    for ShapeList<SeriesSource<'source>, PointSource<'source>>
{
    fn from(value: ShapeList<&'source Timeseries, &'source Metric>) -> Self {
        Self {
            shapes: value.shapes.into_iter().map(Shape::into).collect(),
            source: value.source.into(),
        }
    }
}

#[derive(Serialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum Shape<P> {
    Area(Area<P>),
    Line(Line<P>),
    Point(Point<P>),
    Rectangle(Rectangle<P>),
}

impl<'source> From<Shape<&'source Metric>> for Shape<PointSource<'source>> {
    fn from(other: Shape<&'source Metric>) -> Self {
        match other {
            Shape::Area(area) => Shape::Area(area.into()),
            Shape::Line(line) => Shape::Line(line.into()),
            Shape::Point(point) => Shape::Point(point.into()),
            Shape::Rectangle(rectangle) => Shape::Rectangle(rectangle.into()),
        }
    }
}

/// An area to be drawn between two lines that share their X coordinates.
///
/// Area points move from left to right.
#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Area<P> {
    pub points: Vec<AreaPoint<P>>,
}

impl<'source> From<Area<&'source Metric>> for Area<PointSource<'source>> {
    fn from(value: Area<&'source Metric>) -> Self {
        Self {
            points: value.points.into_iter().map(AreaPoint::into).collect(),
        }
    }
}

/// A single data point in an area shape.
#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct AreaPoint<P> {
    /// X coordinate between 0.0 and 1.0.
    pub x: f64,

    /// Y coordinate between 0.0 and 1.0 for the bottom of the area.
    pub y_min: f64,

    /// Y coordinate between 0.0 and 1.0 for the top of the area.
    pub y_max: f64,

    /// The source this point was generated from.
    ///
    /// This would be a [Metric] if the chart was generated from [Timeseries].
    pub source: P,
}

impl<'source> From<AreaPoint<&'source Metric>> for AreaPoint<PointSource<'source>> {
    fn from(value: AreaPoint<&'source Metric>) -> Self {
        Self {
            x: value.x,
            y_min: value.y_min,
            y_max: value.y_max,
            source: value.source.into(),
        }
    }
}

/// A line to be drawn between two or more points.
#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Line<P> {
    pub points: Vec<Point<P>>,
}

impl<'source> From<Line<&'source Metric>> for Line<PointSource<'source>> {
    fn from(value: Line<&'source Metric>) -> Self {
        Self {
            points: value.points.into_iter().map(Point::into).collect(),
        }
    }
}

/// A single point in the chart.
///
/// Points can be rendered independently as a dot, or can be used to draw lines
/// between them.
#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Point<P> {
    /// X coordinate between 0.0 and 1.0.
    pub x: f64,

    /// Y coordinate between 0.0 and 1.0.
    pub y: f64,

    /// The source this point was generated from.
    ///
    /// This would be a [Metric] if the chart was generated from [Timeseries].
    pub source: P,
}

impl<'source> From<Point<&'source Metric>> for Point<PointSource<'source>> {
    fn from(value: Point<&'source Metric>) -> Self {
        Self {
            x: value.x,
            y: value.y,
            source: value.source.into(),
        }
    }
}

/// A rectangle to be rendered inside the chart.
#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Rectangle<P> {
    pub x: f64,
    pub y: f64,
    pub width: f64,
    pub height: f64,

    /// The source this rectangle was generated from.
    ///
    /// This would be a [Metric] if the chart was generated from [Timeseries].
    pub source: P,
}

impl<'source> From<Rectangle<&'source Metric>> for Rectangle<PointSource<'source>> {
    fn from(value: Rectangle<&'source Metric>) -> Self {
        Self {
            x: value.x,
            y: value.y,
            width: value.width,
            height: value.height,
            source: value.source.into(),
        }
    }
}

pub(crate) type StackedChartBuckets = Buckets<StackedChartBucketValue>;

#[derive(Clone, Copy, Default)]
pub(crate) struct StackedChartBucketValue {
    /// Used to keep track of how much a bucket is "filled" while calculating
    /// the area shapes.
    pub current_y: f64,

    /// The sum of all values in the bucket.
    pub total: f64,
}
