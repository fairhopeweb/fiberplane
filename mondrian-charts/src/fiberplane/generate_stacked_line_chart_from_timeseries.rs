use super::utils::*;
use crate::fiberplane::{Metric, StackedChartBuckets, Timeseries, TimeseriesSourceData};
use crate::types::{Area, AreaPoint, Axis, MondrianChart, Shape, ShapeList};

pub(crate) fn generate_stacked_line_chart_from_timeseries<'source>(
    input: TimeseriesSourceData<'source, '_, '_>,
) -> MondrianChart<&'source Timeseries, &'source Metric> {
    let BucketsAndAxes {
        mut buckets,
        is_percentage,
        mut x_axis,
        mut y_axis,
    } = calculate_buckets_and_axes_for_stacked_chart(&input);

    for &value in input.additional_values {
        y_axis = y_axis.extend_with_value(value);
    }

    let interval = calculate_smallest_time_interval(&buckets);
    if let Some(interval) = interval {
        attach_suggestions_to_x_axis(&mut x_axis, &buckets, interval);
    }

    let mut args = BarArgs {
        buckets: &mut buckets,
        is_percentage,
        x_axis: &x_axis,
        y_axis: &y_axis,
    };

    let shape_lists: Vec<_> = input
        .timeseries_data
        .iter()
        .map(|timeseries| ShapeList {
            shapes: if timeseries.visible {
                create_shapes(&timeseries.metrics, interval, &mut args)
            } else {
                Vec::new()
            },
            source: *timeseries,
        })
        .collect();

    MondrianChart {
        shape_lists,
        x_axis,
        y_axis,
    }
}

struct BarArgs<'axes, 'buckets> {
    buckets: &'buckets mut StackedChartBuckets,
    is_percentage: bool,
    x_axis: &'axes Axis,
    y_axis: &'axes Axis,
}

fn create_shapes<'source>(
    metrics: &'source [Metric],
    interval: Option<f64>,
    args: &mut BarArgs,
) -> Vec<Shape<&'source Metric>> {
    split_into_continuous_lines(metrics, interval)
        .iter()
        .map(|line| {
            Shape::Area(Area {
                area_gradient_shown: None,
                points: line
                    .iter()
                    .filter_map(|metric| create_point_for_metric(metric, args))
                    .collect(),
            })
        })
        .collect()
}

fn create_point_for_metric<'source>(
    metric: &'source Metric,
    args: &mut BarArgs,
) -> Option<AreaPoint<&'source Metric>> {
    let bucket_value = args.buckets.get_mut(&metric.time)?;

    let time = get_time_from_timestamp(metric.time);
    let value = if args.is_percentage {
        metric.value / bucket_value.total
    } else {
        metric.value
    };

    let y_min = bucket_value.current_y;
    let y_max = y_min + normalize_along_linear_axis(value, args.y_axis);
    bucket_value.current_y = y_max;

    Some(AreaPoint {
        x: normalize_along_linear_axis(time, args.x_axis),
        y_min,
        y_max,
        source: metric,
    })
}
