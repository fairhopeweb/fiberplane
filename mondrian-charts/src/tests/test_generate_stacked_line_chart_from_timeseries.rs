use crate::generate_stacked_line_chart_from_timeseries;
use crate::types::{
    GraphType, Metric, StackingType, TimeRange, Timeseries, TimeseriesSourceData, Timestamp,
};
use std::collections::BTreeMap;
use std::str::FromStr;

#[test]
fn test_generate_stacked_line_chart_from_timeseries() {
    let timeseries_a = Timeseries::builder()
        .name("dummy_data")
        .metrics(vec![
            get_metric_at_minute(0, 10.),
            get_metric_at_minute(1, 15.),
            get_metric_at_minute(2, 20.),
            get_metric_at_minute(4, 25.),
            get_metric_at_minute(5, 30.),
            get_metric_at_minute(6, 25.),
            get_metric_at_minute(7, 20.),
            get_metric_at_minute(8, 15.),
            get_metric_at_minute(9, 10.),
        ])
        .labels(BTreeMap::from([("timeseries".to_owned(), "A".to_owned())]))
        .visible(true)
        .build();

    let timeseries_b = Timeseries::builder()
        .name("dummy_data")
        .metrics(vec![
            get_metric_at_minute(0, 40.),
            get_metric_at_minute(1, 30.),
            get_metric_at_minute(2, 20.),
            get_metric_at_minute(3, 10.),
            get_metric_at_minute(4, 20.),
            get_metric_at_minute(5, 30.),
            get_metric_at_minute(6, 40.),
            get_metric_at_minute(7, 50.),
            get_metric_at_minute(8, 60.),
            get_metric_at_minute(9, 70.),
        ])
        .labels(BTreeMap::from([("timeseries".to_owned(), "B".to_owned())]))
        .visible(true)
        .build();

    let chart = generate_stacked_line_chart_from_timeseries(TimeseriesSourceData {
        graph_type: GraphType::Line,
        stacking_type: StackingType::Stacked,
        time_range: TimeRange {
            from: get_date_at_minute(0),
            to: get_date_at_minute(10),
        },
        timeseries_data: &[&timeseries_a, &timeseries_b],
    });

    insta::assert_yaml_snapshot!(chart);
}

fn get_metric_at_minute(minute: u8, value: f64) -> Metric {
    Metric::builder()
        .time(get_date_at_minute(minute))
        .value(value)
        .build()
}

fn get_date_at_minute(minute: u8) -> Timestamp {
    Timestamp::from_str(&format!("2023-07-18T16:{minute:02}:00.000Z")).unwrap()
}
