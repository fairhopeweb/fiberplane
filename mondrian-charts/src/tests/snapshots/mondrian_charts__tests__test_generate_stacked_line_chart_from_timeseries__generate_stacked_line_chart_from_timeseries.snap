---
source: mondrian-charts/src/tests/test_generate_stacked_line_chart_from_timeseries.rs
expression: chart
---
xAxis:
  minValue: 1689696000000
  maxValue: 1689696600000
  tickSuggestions:
    - 1689696000000
    - 1689696060000
    - 1689696120000
    - 1689696180000
    - 1689696240000
    - 1689696300000
    - 1689696360000
    - 1689696420000
    - 1689696480000
    - 1689696540000
yAxis:
  minValue: 0
  maxValue: 80
  tickSuggestions: ~
shapeLists:
  - shapes:
      - type: area
        points:
          - x: 0
            yMin: 0
            yMax: 0.125
            source:
              time: "2023-07-18T16:00:00Z"
              attributes: {}
              resource: {}
              value: 10
          - x: 0.1
            yMin: 0
            yMax: 0.1875
            source:
              time: "2023-07-18T16:01:00Z"
              attributes: {}
              resource: {}
              value: 15
          - x: 0.2
            yMin: 0
            yMax: 0.25
            source:
              time: "2023-07-18T16:02:00Z"
              attributes: {}
              resource: {}
              value: 20
      - type: area
        points:
          - x: 0.4
            yMin: 0
            yMax: 0.3125
            source:
              time: "2023-07-18T16:04:00Z"
              attributes: {}
              resource: {}
              value: 25
          - x: 0.5
            yMin: 0
            yMax: 0.375
            source:
              time: "2023-07-18T16:05:00Z"
              attributes: {}
              resource: {}
              value: 30
          - x: 0.6
            yMin: 0
            yMax: 0.3125
            source:
              time: "2023-07-18T16:06:00Z"
              attributes: {}
              resource: {}
              value: 25
          - x: 0.7
            yMin: 0
            yMax: 0.25
            source:
              time: "2023-07-18T16:07:00Z"
              attributes: {}
              resource: {}
              value: 20
          - x: 0.8
            yMin: 0
            yMax: 0.1875
            source:
              time: "2023-07-18T16:08:00Z"
              attributes: {}
              resource: {}
              value: 15
          - x: 0.9
            yMin: 0
            yMax: 0.125
            source:
              time: "2023-07-18T16:09:00Z"
              attributes: {}
              resource: {}
              value: 10
    source:
      name: dummy_data
      labels:
        timeseries: A
      metrics:
        - time: "2023-07-18T16:00:00Z"
          attributes: {}
          resource: {}
          value: 10
        - time: "2023-07-18T16:01:00Z"
          attributes: {}
          resource: {}
          value: 15
        - time: "2023-07-18T16:02:00Z"
          attributes: {}
          resource: {}
          value: 20
        - time: "2023-07-18T16:04:00Z"
          attributes: {}
          resource: {}
          value: 25
        - time: "2023-07-18T16:05:00Z"
          attributes: {}
          resource: {}
          value: 30
        - time: "2023-07-18T16:06:00Z"
          attributes: {}
          resource: {}
          value: 25
        - time: "2023-07-18T16:07:00Z"
          attributes: {}
          resource: {}
          value: 20
        - time: "2023-07-18T16:08:00Z"
          attributes: {}
          resource: {}
          value: 15
        - time: "2023-07-18T16:09:00Z"
          attributes: {}
          resource: {}
          value: 10
      attributes: {}
      resource: {}
      visible: true
  - shapes:
      - type: area
        points:
          - x: 0
            yMin: 0.125
            yMax: 0.625
            source:
              time: "2023-07-18T16:00:00Z"
              attributes: {}
              resource: {}
              value: 40
          - x: 0.1
            yMin: 0.1875
            yMax: 0.5625
            source:
              time: "2023-07-18T16:01:00Z"
              attributes: {}
              resource: {}
              value: 30
          - x: 0.2
            yMin: 0.25
            yMax: 0.5
            source:
              time: "2023-07-18T16:02:00Z"
              attributes: {}
              resource: {}
              value: 20
          - x: 0.3
            yMin: 0
            yMax: 0.125
            source:
              time: "2023-07-18T16:03:00Z"
              attributes: {}
              resource: {}
              value: 10
          - x: 0.4
            yMin: 0.3125
            yMax: 0.5625
            source:
              time: "2023-07-18T16:04:00Z"
              attributes: {}
              resource: {}
              value: 20
          - x: 0.5
            yMin: 0.375
            yMax: 0.75
            source:
              time: "2023-07-18T16:05:00Z"
              attributes: {}
              resource: {}
              value: 30
          - x: 0.6
            yMin: 0.3125
            yMax: 0.8125
            source:
              time: "2023-07-18T16:06:00Z"
              attributes: {}
              resource: {}
              value: 40
          - x: 0.7
            yMin: 0.25
            yMax: 0.875
            source:
              time: "2023-07-18T16:07:00Z"
              attributes: {}
              resource: {}
              value: 50
          - x: 0.8
            yMin: 0.1875
            yMax: 0.9375
            source:
              time: "2023-07-18T16:08:00Z"
              attributes: {}
              resource: {}
              value: 60
          - x: 0.9
            yMin: 0.125
            yMax: 1
            source:
              time: "2023-07-18T16:09:00Z"
              attributes: {}
              resource: {}
              value: 70
    source:
      name: dummy_data
      labels:
        timeseries: B
      metrics:
        - time: "2023-07-18T16:00:00Z"
          attributes: {}
          resource: {}
          value: 40
        - time: "2023-07-18T16:01:00Z"
          attributes: {}
          resource: {}
          value: 30
        - time: "2023-07-18T16:02:00Z"
          attributes: {}
          resource: {}
          value: 20
        - time: "2023-07-18T16:03:00Z"
          attributes: {}
          resource: {}
          value: 10
        - time: "2023-07-18T16:04:00Z"
          attributes: {}
          resource: {}
          value: 20
        - time: "2023-07-18T16:05:00Z"
          attributes: {}
          resource: {}
          value: 30
        - time: "2023-07-18T16:06:00Z"
          attributes: {}
          resource: {}
          value: 40
        - time: "2023-07-18T16:07:00Z"
          attributes: {}
          resource: {}
          value: 50
        - time: "2023-07-18T16:08:00Z"
          attributes: {}
          resource: {}
          value: 60
        - time: "2023-07-18T16:09:00Z"
          attributes: {}
          resource: {}
          value: 70
      attributes: {}
      resource: {}
      visible: true

